### PR TITLE
OBPIH 2519.  Dashboard: fill rate (Historical)

### PR DESCRIPTION
**Pull request by julienreliefapps**
As the location filter contains more things, we did it a bit in a different way. If this proposition is accepted, we will improve the responsiveness of this one. Else, we could think about making a pop up for this kind of "big" filter

### CODE REVIEW

**Comment by awalkowiak** [comment]
@julienreliefapps seems like two failing tests (InventoryItemControllerTests.test_showRecordInventory and ProductTests.toJson) were introduced by my changes. I fix those two.

**Comment by julienreliefapps** [comment]
@awalkowiak thank you !

**Comment by jmiranda** [comment]
@awalkowiak I think @julienreliefapps based this feature branch off a version of develop before I fixed the issue. @julienreliefapps You can try to rebase it to develop.

**Comment by julienreliefapps** [comment]
@jmiranda  I have created the branch from develop just this morning, did you make this change today ?

**Comment by awalkowiak** [comment]
@jmiranda I have fresh develop locally and these tests are still failing for me. It is because of https://github.com/openboxes/openboxes/blob/develop/grails-app/domain/org/pih/warehouse/product/Product.groovy#L589. Were you fixing this issue already? If no, then I am trying to fix this.

**Comment by jmiranda** [comment]
Ah yeah I fixed a similar issue a few days ago. https://github.com/openboxes/openboxes/commit/3b48784f48c8e495b6cf3d6c6fd940c56416e8a7

Unit tests don't have access to any of the grailsApplication junk so you either have to mock it or @ignore the failing tests.

**Comment by jmiranda** [comment]
@awalkowiak Review this when you get a chance and if it looks ok please merge so that Katarzyna can test it and move it to UAT.

File: grails-app/controllers/org/pih/warehouse/api/ApiController.groovy [hunk 1]
```diff
@@ -116,6 +117,7 @@ class ApiController {
         User user = User.get(session?.user?.id)
         Location location = Location.get(session.warehouse?.id)
         String highestRole = user.getHighestRole(location)
+        def allLocations = locationService.getAllLocations()
```
Comment (awalkowiak) [inline grails-app/controllers/org/pih/warehouse/api/ApiController.groovy:120]: Do we really want to have it here? And should it be all locations? With this locally getAppContext takes 2s to process while without this it took for me less than 200ms. On obdev getAppContext right now takes a bit less than 600 ms (from which major part was just initial connection). I don't like the fact that getAppContext is overloaded with so much info right now. I think it should be done in a separate endpoint. In fact, there already is `/api/locations`.

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy [hunk 2]
```diff
@@ -94,38 +94,89 @@ class IndicatorDataService {
...
+            GROUP BY MONTH(fr.transaction_date), YEAR(fr.transaction_date)
+            """, [
+                'monthEnd'    : monthEnd,
+                'monthBegin'  : monthBegin,
+                'destination' : destination.id,
+                'origin'      : location.id,   
+            ]);
+
+            requestLinesSubmitted[0] == null ? requestLinesSubmittedResult.push(0) : requestLinesSubmittedResult.push(requestLinesSubmitted[0][0])
+println(destination.id)
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:146]: Is this required?

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy [hunk 3]
```diff
@@ -94,38 +94,89 @@ class IndicatorDataService {
...
+println(location.id)
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:147]: Is this required?

File: package-lock.json [hunk 4]
```diff
@@ -1311,7 +1311,7 @@
     },
     "ansi-escapes": {
       "version": "3.1.0",
-      "resolved": "https://registry.npmjs.org/ansi-escapes/-/ansi-escapes-3.1.0.tgz",
+      "resolved": "http://registry.npmjs.org/ansi-escapes/-/ansi-escapes-3.1.0.tgz",
```
Comment (awalkowiak) [inline package-lock.json:1314]: How those changes happened to be here? I don't see any package.json updates.

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy [hunk 5]
```diff
@@ -94,38 +94,89 @@ class IndicatorDataService {
...
             monthEnd.set(month: today.month - i + 1, date: 1)
+            
+            String monthLabel = new java.text.DateFormatSymbols().months[monthBegin.month]
 
-            def query1 = Requisition.executeQuery("""select count(*) from RequisitionItem where dateCreated >= ? and dateCreated < ?""", [monthBegin, monthEnd]);
+            listLabel.push("${monthLabel} ${monthBegin.year + 1900}")
 
-            def query2 = Requisition.executeQuery("""select count(*) from RequisitionItem where dateCreated >= ? and dateCreated < ? and quantityCanceled > 0 and (cancelReasonCode = 'STOCKOUT' or cancelReasonCode = 'LOW_STOCK' or cancelReasonCode = 'COULD_NOT_LOCATE')""", [monthBegin, monthEnd]);
-            String monthLabel = new java.text.DateFormatSymbols().months[monthBegin.month]
+            def averageFillRate = dataService.executeQuery("""
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:121]: Locally I am getting here Internal server error with message:`SELECT command denied to user 'openboxes_test'@'localhost' for table 'fill_rate'`. Should there be a migration with this table?

**Comment by awalkowiak** [comment]
btw @julienreliefapps if you rebase this to develop all tests that are failing because of me will be fixed.

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy (hunk 2 above)
Comment (julienreliefapps) [reply grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:146]: Absolutely not, I forgot these 2 ones

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy (hunk 5 above)
Comment (julienreliefapps) [reply grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:121]: Yes, the migration hasn't been done yet, you have to create the view locally : 

```
create or replace view fill_rate as 
SELECT 
r.id, 
r.status, 
ri.quantity, 
ri.quantity_canceled, 
ri.cancel_reason_code, 
ri.requisition_item_type, 
product.id as product_id,
product.product_code,
product.name as product_name,
t.transaction_date, month(t.transaction_date) as month, 
year(t.transaction_date) as year,
origin.id as origin_id, 
origin.name as origin_name,
destination.id as destination_id,
destination.name as destination_name,
case
when 
il.status = 'FORMULARY' then true
else false
end as formulary,
case
when 
ri.quantity_canceled > 0 
and ri.cancel_reason_code in ('STOCKOUT','LOW_STOCK','COULD_NOT_LOCATE') then 0 
else 1
end as fill_rate
FROM requisition as r 
join requisition_item ri on r.id = ri.requisition_id
join location as origin on origin.id = r.origin_id
join location as destination on destination.id = r.destination_id
join product as product on product.id = ri.product_id
left outer join transaction as t on t.requisition_id = r.id
left outer join inventory_level as il on ( il.inventory_id = origin.inventory_id and il.product_id = product.id )
where ri.requisition_item_type = 'ORIGINAL'
and r.status = 'ISSUED';
```

File: package-lock.json (hunk 4 above)
Comment (julienreliefapps) [reply package-lock.json:1314]: @awalkowiak  I don't really know for this one, I got that pulling develop and doing npm install, I will have a look at the history of develop

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy (hunk 3 above)
Comment (julienreliefapps) [reply grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:147]: Absolutely not, I forgot these 2 ones

File: grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy (hunk 5 above)
Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:121]: ok

Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy:121]: Then I am adding 'needs-migration' label
cc @jmiranda

File: grails-app/controllers/org/pih/warehouse/api/ApiController.groovy [hunk 6]
```diff
@@ -26,6 +26,7 @@ class ApiController {
     def dataSource
     def userService
     def localizationService
+    def locationService
```
Comment (awalkowiak) [inline grails-app/controllers/org/pih/warehouse/api/ApiController.groovy:29]: Leftover

**Comment by jmiranda** [comment]
@awalkowiak Thanks for catching the location API issue. That would have been bad if not caught.

File: grails-app/controllers/org/pih/warehouse/api/ApiController.groovy (hunk 6 above)
Comment (julienreliefapps) [reply grails-app/controllers/org/pih/warehouse/api/ApiController.groovy:29]: ah yes it's true, I remove that now

**Comment by julienreliefapps** [comment]
@jmiranda don't merge it right now, with your video, I just detected an issue, I don't know if it was before but I will fix it right now

**Comment by jmiranda** [comment]
@julienreliefapps Ok. And if you could fix the fill rate percentage axis to always go from 0-100% that would be great. Right now sometimes it shows 90% as the max.

**Pull request merged**